### PR TITLE
Fixes related to handling unusual double values (+/- infinity and NaN).

### PR DESCRIPTION
**Pull request by arbimo**
Just a few fixes related to edges cases involving infinities and NaN.
One common error was the use of `Constant(d)` to construct a `Real` from a `Double`, this would throw a `NumberFormatException` on NaN and infinities when converting to `BigDecimal`. Using `Real(d)` should handle infinities and throw an arithmetic error on NaN.

### CODE REVIEW

File: rainier-tests/src/test/scala/com/stripe/rainier/compute/RealTest.scala [hunk 1]
```diff
@@ -129,7 +139,15 @@ class RealTest extends FunSuite {
     Real.gt(x, 0, x * x, x + 1)
   }
 
-  run("normal") { x =>
+  run("normal", defined = !_.isPosInfinity) { x =>
+    // FIXME: letting this be defined at +infinity results in the following error:
+    //     -Infinity did not equal NaN [c/ev, n=Infinity] (RealTest.scala:67)
+    // I.e. two different results from constant folding and evaluation.
+    // same applies to the "normal sum" test case below
+    Normal(x, 1).logDensity(Real(1d))
```
Comment (arbimo) [inline rainier-tests/src/test/scala/com/stripe/rainier/compute/RealTest.scala:147]: This is a potential problem identified by the additional tests where constant folding and evaluation yield different results.
This is really a edge case but might a symptom of a more general problem.

The problematic case is excluded from the tests since I won't attempt to fix it in this PR but it probably warrants a separate bug report.

**Comment by avi-stripe** [comment]
Thank you!
